### ISSUE TITLE
Many subscriptions take a long time

### ISSUE THREAD

**Issue by dobermai** [eclipse-paho/paho.mqtt.java#167]
(from https://bugs.eclipse.org/bugs/show_bug.cgi?id=487650)

When subscribing to many topics at once, it takes significant time until the subscribe() method returns. After further investigation it turned out this is broker independent. I attached a test which shows the issue. The test shows that 8000 topics need approx. 40 seconds (on my machine) to subscribe to. To verify this is not a broker problem, we created a very basic MQTT client and subscribed to multiple thousand topics and it took a few milliseconds (end-to-end) to subscribe to that many topics.

While most of MQTT use cases don't require to have multiple hundreds of subscriptions, we have some customer use cases where even thousands of single subscriptions per client are possible and this is a potential blocker for them to use Paho.

import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
import org.junit.Test;

```
/**
 * @author Dominik Obermaier
 */
public class SubscriptionTest {

    @Test
    public void test_many_subscriptions() throws Exception {

        final MqttClient client = new MqttClient("tcp://broker.hivemq.com:1883", MqttClient.generateClientId(), new MemoryPersistence());

        client.connect();


        final String[] topicFilters = new String[8000];
        for (int i = 0; i < 8000; i++) {
            topicFilters[i] = "my_topic" + i;
        }
        final long now = System.currentTimeMillis();
        System.out.println("Starting to subscribe");
        client.subscribe(topicFilters);
        System.out.println("Took " + (System.currentTimeMillis() - now) + " milliseconds to subscribe");

        client.disconnect();

    }
}
```

**Comment by dobermai** [comment]
Since the Neon release is around the corner, I was curious if this is still scheduled for 1.2.0?

**Comment by jpwsutton** [comment]
I've been testing this every now and again and am having trouble actually verifying if this is a bug with the client itself. I ran this code against a local mosquitto broker (Version 1.4.8) and it took 2550 milliseconds to make the 8000 subscriptions. Wireshark Capture Below:

![image](https://cloud.githubusercontent.com/assets/1068763/15668883/704e0766-2715-11e6-915a-7115851d391d.png)

Then, when tested against iot.eclipse.org (another mosquitto server) the delay was much greater: 58869 ms.

![image](https://cloud.githubusercontent.com/assets/1068763/15669185/dd40d8de-2716-11e6-84c2-7a08b507b97a.png)

broker.mqttdashboard.com was a lot faster (3910 ms):

![image](https://cloud.githubusercontent.com/assets/1068763/15669249/485d5d86-2717-11e6-9817-d221f3cc03ed.png)

What's interesting is that the delay doesn't seem to be in the client, but on the server side, this may be something to do with the load that the server is under, or the number of subscriptions currently in memory possibly?

When you wrote your test client did you make a bulk subscription like this code does, or do you make individual subscription requests?

**Comment by dobermai** [comment]
I just executed the tests again and I was not able to reproduce this issue again with 8000 subscriptions with the 1.0.2 version. I tried multiple brokers and all of them seem to work fine (except iot.eclipse.org).

However, I increased the topics to 30000 topics in the same SUBSCRIBE packet (instead of 8000) and I see the following with a locally installed mosquitto 1.4.8:

<img width="1044" alt="screen shot 2016-05-31 at 22 15 07" src="https://cloud.githubusercontent.com/assets/684310/15689026/527d9d42-277d-11e6-8a01-81faf389c0ce.png">

As you can see, the SUBSCRIBE message takes over 20 seconds just to be sent over the wire. So I guess this is a problem on the client side when creating huge packets.

I double check with another local broker (in this case HiveMQ) and we still see that the SUBSCRIBE message takes on my machine 20 seconds to send out (the SUBACK is faster, though):

<img width="1048" alt="screen shot 2016-05-31 at 22 18 22" src="https://cloud.githubusercontent.com/assets/684310/15689113/bf4188f8-277d-11e6-8888-fdbd1c8d9d6e.png">

If it helps reproducing the issue, this is the exact code I run with Paho 1.0.2:

```
   @Test
    public void test_many_subscriptions() throws Exception {

        final MqttClient client = new MqttClient("tcp://localhost:1883", MqttClient.generateClientId(), new MemoryPersistence());

        client.connect();


        final String[] topicFilters = new String[30000];
        for (int i = 0; i < 30000; i++) {
            topicFilters[i] = "my_topic" + i;
        }
        final long now = System.currentTimeMillis();
        System.out.println("Starting to subscribe");
        client.subscribe(topicFilters);
        System.out.println("Took " + (System.currentTimeMillis() - now) + " milliseconds to subscribe");

        client.disconnect();
    }
```

**Comment by dobermai** [comment]
Just an additional note:

I recreated the scenario with the Fusesource MQTT Client and the issue doesnt appear to happen with this client. Here is what is happening on the wire:

<img width="1051" alt="screen shot 2016-05-31 at 22 25 48" src="https://cloud.githubusercontent.com/assets/684310/15689319/bcaa0d08-277e-11e6-96bf-df721aee0a0f.png">

And for reference this is the code I used:

```
 @Test
    public void test_many_subscriptions_fusesource_client() throws Exception {

        final MQTT mqtt = new MQTT();
        final BlockingConnection blockingConnection = mqtt.blockingConnection();

        blockingConnection.connect();

        final Topic[] topicFilters = new Topic[30000];
        for (int i = 0; i < 30000; i++) {
            topicFilters[i] = new Topic("my_topic" + i, QoS.AT_MOST_ONCE);
        }
        final long now = System.currentTimeMillis();
        System.out.println("Starting to subscribe");
        blockingConnection.subscribe(topicFilters);
        System.out.println("Took " + (System.currentTimeMillis() - now) + " milliseconds to subscribe");

        blockingConnection.disconnect();
    }
```

**Comment by dobermai** [comment]
It turns out, that just by commenting out the trace logging fixes the issue.

The code in question is in class MqttAsyncClient.java

The following modification does the trick. Of course this is just for demonstration purposes, the real fix should be:
1. Do a check if tracelog is activated, otherwise skip the code path
2. Use a StringBuilder instead of String concatenation

```
public IMqttToken subscribe(String[] topicFilters, int[] qos, Object userContext, IMqttActionListener callback) throws MqttException {
        final String methodName = "subscribe";

        if (topicFilters.length != qos.length) {
            throw new IllegalArgumentException();
        }

        // remove any message handlers for individual topics
        for (int i = 0; i < topicFilters.length; ++i) {
            this.comms.removeMessageListener(topicFilters[i]);
        }

//        String subs = "";
//        for (int i = 0; i < topicFilters.length; i++) {
//            if (i > 0) {
//                subs += ", ";
//            }
//            subs += "topic=" + topicFilters[i] + " qos=" + qos[i];
//
//            //Check if the topic filter is valid before subscribing
//            MqttTopic.validate(topicFilters[i], true/*allow wildcards*/);
//        }
//        //@TRACE 106=Subscribe topicFilter={0} userContext={1} callback={2}
//        log.fine(CLASS_NAME, methodName, "106", new Object[]{subs, userContext, callback});

        MqttToken token = new MqttToken(getClientId());
        token.setActionCallback(callback);
        token.setUserContext(userContext);
        token.internalTok.setTopics(topicFilters);

        MqttSubscribe register = new MqttSubscribe(topicFilters, qos);

        comms.sendNoWait(register, token);
        //@TRACE 109=<
        log.fine(CLASS_NAME, methodName, "109");

        return token;
    }
```

**Comment by dobermai** [comment]
I just submitted a fix for that issue. The fix should be compatible with Java 1.4.2. The Pull Request that fixes the issue is #243
